### ISSUE TITLE
[OPS] optimize token usage

### ISSUE THREAD

**Issue by amirfz** [Aggregate-Intellect/sherpa#133]
**PROBLEM**
currently our prompts and llm calls are done in an adhoc manner. this results in redundant or unnecessary text in context or generated steps, or similar issues which in turn increases the cost of operating the system

**SOLUTION**
review all prompts, and LLM calls and identify opportunities for reducing token usage

**ALTERNATIVES**
n/a

**OUTCOME**
the desired outcome is a list of places where token usage can be reduced without impacting system performance and issues created to address them

**ADDITIONAL INFO**
this is related to #98 and #18 that are addressed separately 

**EXAMPLE**
The following output has several related text generations 
```
#verbose message:  [{'Step': '0/5', 'reply': {'thoughts': {'text': 'A JSON schema is a file that defines the structure, data types, and validation rules for a JSON document. It provides a way to describe the expected format of JSON data.', 'speak': 'A JSON schema is a file that defines the structure, data types, and validation rules for a JSON document. It provides a way to describe the expected format of JSON data.'}, 'command': {'name': 'finish', 'args': {'response': 'A JSON schema is a file that defines the structure, data types, and validation rules for a JSON document. It provides a way to describe the expected format of JSON data.'}}}}, {'Step': '6/5', 'reply': 'A JSON schema is a file that defines the structure, data types, and validation rules for a JSON document. It provides a way to describe the expected format of JSON data. You can learn more about JSON schemas here: [JSON Schema](https://json-schema.org/)'}]
```

**Comment by amirfz** [comment]
@oshoma @20001LastOrder @bharatji30 I'm checking this out and I had written more details than I thought. could you let me know what other details are useful to make this more concrete?

**Comment by 20001LastOrder** [comment]
Closed as  #98 and #18 are completed
